### PR TITLE
Change the search of local function to non-recursive version

### PR DESCRIPTION
**Pull request by Cosifne**
Potential fix for https://dev.azure.com/devdiv/DevDiv/_workitems/edit/1582837
@ryzngard and @dibarbet investigate this and it looks like caused by the stackoverflow in the recursive function.

### CODE REVIEW

**Review by ryzngard — APPROVED** [review]

**Review by genlu — APPROVED** [review]

**Comment by CyrusNajmabadi** [comment]
Can we have a test for this?

File: src/Workspaces/CSharp/Portable/FindSymbols/CSharpDeclaredSymbolInfoFactoryService.cs [hunk 1]
```diff
@@ -166,21 +167,21 @@ private static void ProcessUsings(List<Dictionary<string, string>> aliasMaps, Sy
             string fullyQualifiedContainerName,
             CancellationToken cancellationToken)
         {
-            AddLocalFunctionInfosRecurse(memberDeclaration);
-
-            return;
-
-            void AddLocalFunctionInfosRecurse(SyntaxNode node)
+            var queue = new Queue<SyntaxNodeOrToken>();
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/FindSymbols/CSharpDeclaredSymbolInfoFactoryService.cs:170]: can we pool these?

**Comment by Cosifne** [comment]
> Can we have a test for this?

Fred and David B have done a great job to get a code sample to reproduce this. And I am preparing the test

**Comment by Cosifne** [comment]
Update for the test:
I try to test this for a while, but stack overflow is not something easy to repro in unit.
And there is another stackover exception crashes the devenv https://sourceroslyn.io/#Microsoft.CodeAnalysis.Features/Wrapping/BinaryExpression/AbstractBinaryExpressionWrapper.cs,f461f8abcebe1530,references
As long as this issue is resolved we should be able to cover this in integration test.
https://github.com/dotnet/roslyn/issues/63349

File: src/Workspaces/CSharp/Portable/FindSymbols/CSharpDeclaredSymbolInfoFactoryService.cs [hunk 2]
```diff
@@ -166,21 +169,22 @@ private static void ProcessUsings(List<Dictionary<string, string>> aliasMaps, Sy
             string fullyQualifiedContainerName,
             CancellationToken cancellationToken)
         {
-            AddLocalFunctionInfosRecurse(memberDeclaration);
-
-            return;
-
-            void AddLocalFunctionInfosRecurse(SyntaxNode node)
+            using var pooledQueue = s_queuePool.GetPooledObject();
```
Comment (genlu) [inline src/Workspaces/CSharp/Portable/FindSymbols/CSharpDeclaredSymbolInfoFactoryService.cs:172]: maybe `using var queue = SharedPools.Default<Queue<SyntaxNodeOrToken>>().GetPooledObject();` ?

**Review by CyrusNajmabadi — APPROVED** [review]

**Comment by dibarbet** [comment]
@Cosifne once https://github.com/dotnet/roslyn/pull/63372 goes in you should be able to cherry-pick this change on top of a revert to the original revert.

**Comment by Cosifne** [comment]
This is checked in the other PR.

**Pull request closed without merging**